src/k2/tso/service/Service.cpp

```diff
@@ -0,0 +1,220 @@
+                    throw std::runtime_error("Unable to set affinity");
+                }
+            }
+            while (_keepRunningPoller.test_and_set()) {
+                GPSClock::poll();
+            }
+        });
+        startFut = startFut.then([] {
+            // notify all workers that the clock is initialized and we can start
+            return k2::AppBase().getDist<TSOService>().invoke_on_all([] (auto& svc) {
+                svc._clockInitialized.set_value();
+            });
+        });
+    }
+
+    return startFut
+        .then([this] {
+            // wait for clock to be initialized
+            return _clockInitialized.get_future();
+        })
+        .then([this] {
+            return _collectWorkerURLs();
+        })
+        .then([this] {
+            // TODO also calibrate local error bound
+            return _cpoRegister();
+        })
+        .then([this] {
+            // register RPC APIs
+            RPC().registerRPCObserver<dto::GetServiceNodeURLsRequest, dto::GetServiceNodeURLsResponse>
+            (dto::Verbs::GET_TSO_SERVICE_NODE_URLS, [this](dto::GetServiceNodeURLsRequest&& request) {
+                return _handleGetServiceNodeURLs(std::move(request));
+            });
+
+            RPC().registerRPCObserver<dto::GetTimestampRequest, dto::GetTimestampResponse>(dto::Verbs::GET_TSO_TIMESTAMP, [this](dto::GetTimestampRequest&& request) {
+                return _handleGetTimestamp(std::move(request));
+            });
+
+            K2LOG_I(log::tsoserver, "TSOService started");
+            return seastar::make_ready_future<>();
+        });
+}
+
+seastar::future<> TSOService::gracefulStop() {
+    K2LOG_I(log::tsoserver, "TSOService stop");
+    if (_clockPoller.joinable()) {
+        K2LOG_I(log::tsoserver, "Stopping clock poller");
+        _keepRunningPoller.clear(); // set the flag to false to signal the thread to exit
+        _clockPoller.join();
+        K2LOG_I(log::tsoserver, "Clock poller stopped");
+    }
+
+    K2LOG_I(log::tsoserver, "done cleaning up");
+
+    return seastar::make_ready_future<>();
+}
+
+seastar::future<> TSOService::_collectWorkerURLs() {
+    K2LOG_I(log::tsoserver, "collecting all worker URLs");
+    std::vector<String> myurls;
+    for (auto& ep : RPC().getServerEndpoints()) {
+        myurls.push_back(ep->url);
+    };
+
+    return AppBase().getDist<TSOService>().invoke_on_all([myurls] (auto& svc) {
+        svc._workersURLs.push_back(myurls);
+    });
+}
+
+seastar::future<> TSOService::_cpoRegister() {
+    // TODO implement CPO-based registration and ID assignment
+    _tsoId = seastar::this_shard_id() + 100000;
+
+    K2LOG_I(log::tsoserver, "Registration with CPO successful with id {}", _tsoId);
+    return seastar::make_ready_future<>();
+}
+
+seastar::future<std::tuple<Status, dto::GetServiceNodeURLsResponse>>
+TSOService::_handleGetServiceNodeURLs(dto::GetServiceNodeURLsRequest&&) {

```

Comment:
What is the purpose of this RPC? Is it just to share the RDMA endpoint? It seems that you need to have the worker url to call this to begin with.